src/common/snippets/include/snippets/op/buffer.hpp

```diff
@@ -14,94 +14,100 @@ namespace op {
-public:
-    OPENVINO_OP("IntermediateMemoryBuffer", "SnippetsOpset", Buffer);
-    IntermediateMemoryBuffer() = default;
-    IntermediateMemoryBuffer(const OutputVector& arguments, size_t allocation_size = utils::get_dynamic_value<size_t>(),
-                             size_t reg_group = 0, size_t cluster_id = 0);
-    IntermediateMemoryBuffer(const ov::Output<ov::Node>& arg, size_t allocation_size = utils::get_dynamic_value<size_t>(),
-                             size_t reg_group = 0, size_t cluster_id = 0);
-
     void validate_and_infer_types() override;
-    std::shared_ptr<Node> clone_with_new_inputs(const OutputVector& new_args) const override;
-};
-
-/**
- * @interface NewMemoryBuffer
- * @brief Represents a new empty memory for allocation with specified shape. It has no parent operations.
- * @ingroup snippets
- *
- */
-class NewMemoryBuffer : public Buffer {
-public:
-    OPENVINO_OP("NewMemoryBuffer", "SnippetsOpset", Buffer);
-    NewMemoryBuffer() = default;
-    NewMemoryBuffer(const ov::Shape& shape, size_t reg_group = 0, size_t cluster_id = 0, ov::element::Type element_type = ov::element::u8);
 
-    void validate_and_infer_types() override;
     std::shared_ptr<Node> clone_with_new_inputs(const OutputVector& new_args) const override;
 
-    void set_element_type(ov::element::Type element_type);
+    size_t get_allocation_size() const { return m_impl->get_allocation_size(); }
 
     class ShapeInfer : public IShapeInferSnippets {
-        ov::Shape m_shape;
+        std::shared_ptr<IShapeInferSnippets> m_impl_shape_infer {nullptr};
     public:
         explicit ShapeInfer(const std::shared_ptr<ov::Node>& n);
         Result infer(const std::vector<VectorDimsRef>& input_shapes) override;
     };
 
 private:
-    ov::Shape m_output_shape;
-    ov::element::Type m_element_type = ov::element::u8;  // u8 - default 1 byte
+    // Base class for implementations of Buffer
+    class BaseImpl;
+    // This constructor is used only in clone_with_new_inputs
+    Buffer(const OutputVector& arguments, std::shared_ptr<BaseImpl> impl);
+
+    const std::shared_ptr<BaseImpl> m_impl {nullptr};
+
+    class BaseImpl {
+    public:
+        BaseImpl() = default;
+        virtual ~BaseImpl() = default;
+        virtual size_t get_allocation_size() const = 0;
+        virtual std::shared_ptr<BaseImpl> clone() const = 0;
+        virtual void validate_and_infer_types(Buffer* buffer) const = 0;
+        virtual bool visit_attributes(AttributeVisitor& visitor) = 0;
+        virtual std::shared_ptr<IShapeInferSnippets> get_shape_infer() const = 0;
+    };
+
+    // IntermediateMemoryImpl represents intermediate memory.
+    // The buffers with this implementation must have source (parents)
+    class IntermediateMemoryImpl : public BaseImpl {
+        class ShapeInfer;
+    public:
+        IntermediateMemoryImpl() = default;
+
+        size_t get_allocation_size() const override { return utils::get_dynamic_value<size_t>(); }
```

Comment:
Could you please remind me why `get_allocation_size` for `IntermediateMemory` always returns dynamic value?

Reply:
`NewMemory` op can be created in data flow where the user must set the allocation size since this node doesn't have parents - cannot take information about allocation size from sources.
`IntermediateMemory` is created only in control_flow where `allocation_size` is calculated in the specific passes. Si I set this value as dynamic.
This method is used only in `BufferExpressionConstructor` when we init attributes. I believe that when we move to all specific-expressions, we will leave only `NewMemoryBuffer` with single method `get_allocation_size` because this method is useless for `IntermediateMemory` ngraph op. But we had to have it for now for conversion to `BufferExpression`.

Reply:
Got it, thanks!